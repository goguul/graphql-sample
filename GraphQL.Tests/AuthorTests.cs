using GraphQL.Authors;
using GraphQL.Books;
using GraphQL.Data;
using HotChocolate;
using HotChocolate.Data;
using HotChocolate.Execution;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;

namespace GraphQL.Tests;

public class AuthorTests
{
    [Fact]
    public async Task Author_Schema_Changed()
    {
        ISchema schema = await new ServiceCollection()
            .AddPooledDbContextFactory<LibraryDbContext>(
                options => options.UseInMemoryDatabase("Data Source=library.db"))
            .AddGraphQL()
            .RegisterDbContext<LibraryDbContext>(DbContextKind.Pooled)
            .AddMutationConventions()
            .AddQueryType()
            .AddMutationType()
            .AddTypeExtension<AuthorQuery>()
            .AddTypeExtension<AuthorMutation>()
            .AddTypeExtension<AuthorType>()
            .AddTypeExtension<AuthorInputType>()
            .AddGlobalObjectIdentification()
            .AddFiltering()
            .AddSorting()
            .AddProjections()
            .BuildSchemaAsync();

        schema.Print().MatchSnapshot();
    }

    [Fact]
    public async Task Add_Author()
    {
        IRequestExecutor requestExecutor = await new ServiceCollection()
            .AddPooledDbContextFactory<LibraryDbContext>(
                options => options.UseInMemoryDatabase("Data Source=library.db"))
            .AddGraphQL()
            .RegisterDbContext<LibraryDbContext>(DbContextKind.Pooled)
            .AddMutationConventions()
            .AddQueryType()
            .AddMutationType()
            .AddTypeExtension<BookType>()
            .AddTypeExtension<AuthorQuery>()
            .AddTypeExtension<AuthorMutation>()
            .AddTypeExtension<AuthorType>()
            .AddTypeExtension<AuthorInputType>()
            .AddGlobalObjectIdentification()
            .AddFiltering()
            .AddSorting()
            .AddProjections()
            .BuildRequestExecutorAsync();

        IExecutionResult executionResult = await requestExecutor.ExecuteAsync(@"
            mutation AddAuthor {
                addAuthor(
                    input: {
                        Name: ""Test Author""
                    }) {
                        Id,
                        Name
                    }
                )
            }
        ");

        executionResult.ToJson().MatchSnapshot();
    }
}