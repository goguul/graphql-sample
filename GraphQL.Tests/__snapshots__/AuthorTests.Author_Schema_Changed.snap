schema {
  query: Query
  mutation: Mutation
}

"The node interface is implemented by entities that have a global unique identifier."
interface Node {
  id: ID!
}

type AddAuthorPayload {
  author: Author
}

type Author implements Node {
  id: ID!
  name: String!
  books: [Book!]!
}

"A connection to a list of items."
type AuthorsConnection {
  "Information to aid in pagination."
  pageInfo: PageInfo!
  "A list of edges."
  edges: [AuthorsEdge!]
  "A flattened list of the nodes."
  nodes: [Author!]
}

"An edge in a connection."
type AuthorsEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Author!
}

type Book {
  id: ID!
  title: String!
  authorId: ID!
  author: Author!
}

type Mutation {
  addAuthor(input: AddAuthorInput!): AddAuthorPayload!
}

"Information about pagination in a connection."
type PageInfo {
  "Indicates whether more edges exist following the set defined by the clients arguments."
  hasNextPage: Boolean!
  "Indicates whether more edges exist prior the set defined by the clients arguments."
  hasPreviousPage: Boolean!
  "When paginating backwards, the cursor to continue."
  startCursor: String
  "When paginating forwards, the cursor to continue."
  endCursor: String
}

type Query {
  "Fetches an object given its ID."
  node("ID of the object." id: ID!): Node
  "Lookup nodes by a list of IDs."
  nodes("The list of node IDs." ids: [ID!]!): [Node]!
  authors("Returns the first _n_ elements from the list." first: Int "Returns the elements in the list that come after the specified cursor." after: String "Returns the last _n_ elements from the list." last: Int "Returns the elements in the list that come before the specified cursor." before: String where: AuthorFilterInput order: [AuthorSortInput!]): AuthorsConnection
  authorById(id: ID!): Author!
}

input AddAuthorInput {
  author: AuthorInput!
}

input AuthorFilterInput {
  and: [AuthorFilterInput!]
  or: [AuthorFilterInput!]
  id: IdOperationFilterInput
  name: StringOperationFilterInput
  books: ListFilterInputTypeOfBookFilterInput
}

input AuthorInput {
  name: String!
}

input AuthorSortInput {
  id: SortEnumType
  name: SortEnumType
}

input BookFilterInput {
  and: [BookFilterInput!]
  or: [BookFilterInput!]
  id: IdOperationFilterInput
  title: StringOperationFilterInput
  authorId: IdOperationFilterInput
  author: AuthorFilterInput
}

input IdOperationFilterInput {
  eq: ID
  neq: ID
  in: [ID]
  nin: [ID]
}

input ListFilterInputTypeOfBookFilterInput {
  all: BookFilterInput
  none: BookFilterInput
  some: BookFilterInput
  any: Boolean
}

input StringOperationFilterInput {
  and: [StringOperationFilterInput!]
  or: [StringOperationFilterInput!]
  eq: String
  neq: String
  contains: String
  ncontains: String
  in: [String]
  nin: [String]
  startsWith: String
  nstartsWith: String
  endsWith: String
  nendsWith: String
}

enum SortEnumType {
  ASC
  DESC
}
